### PR TITLE
Add SoftDelete, HardDelete and UndeleteTree storage methods

### PR DESCRIPTION
**Pull request by codingllama**

### CODE REVIEW

**Comment by codingllama** [comment]
Issue #590

File: storage/testonly/admin_storage_tester.go [hunk 1]
```diff
@@ -202,10 +205,7 @@ func (tester *AdminStorageTester) TestUpdateTree(t *testing.T) {
 	ctx := context.Background()
 	s := tester.NewAdminStorage()
 
-	unrelatedTree, err := createTree(ctx, s, MapTree)
-	if err != nil {
-		t.Fatalf("createTree() = (_, %v), want = (_, nil)", err)
-	}
+	unrelatedTree := makeTreeOrFail(ctx, t.Fatalf, s, spec{Tree: MapTree})
```
Comment (daviddrysdale) [inline storage/testonly/admin_storage_tester.go]: Might read better with the `failFn` as the final argument.

File: storage/testonly/admin_storage_tester.go [hunk 2]
```diff
@@ -579,3 +667,130 @@ func (tester *AdminStorageTester) TestAdminTXClose(t *testing.T) {
...
+	if err != nil {
+		return nil, err
+	}
+	if err := tx.Commit(); err != nil {
+		return nil, err
+	}
+	return tree, nil
+}
+
+func softDeleteTree(ctx context.Context, s storage.AdminStorage, treeID int64) (*trillian.Tree, error) {
```
Comment (daviddrysdale) [inline storage/testonly/admin_storage_tester.go]: might be more consistent to move this function up to just after the 2 soft delete test cases

File: storage/testonly/admin_storage_tester.go [hunk 3]
```diff
@@ -331,188 +331,276 @@ func (tester *AdminStorageTester) TestUpdateTree(t *testing.T) {
...
+	}{
+		{desc: "unknownTree", treeID: 12345, wantCode: errors.NotFound},
+		{desc: "alreadyDeleted", treeID: softDeleted.TreeId, wantCode: errors.FailedPrecondition},
+	}
+	for _, test := range tests {
+		if _, err := softDeleteTree(ctx, s, test.treeID); errors.ErrorCode(err) != test.wantCode {
+			t.Errorf("%v: softDeleteTree() returned err = %v, wantCode = %s", test.desc, err, test.wantCode)
+		}
 	}
 }
```
Comment (daviddrysdale) [inline storage/testonly/admin_storage_tester.go]: I wonder whether it's worth having a `TestFullDelete` case that does the sequence that would occur in practice: create, (maybe) freeze, soft-delete, hard-delete.  

(However, all the individual pieces are already present and checked, so it wouldn't test anything new.)

File: storage/testonly/admin_storage_tester.go (hunk 1 above)
Comment (codingllama) [reply storage/testonly/admin_storage_tester.go]: Done.

File: storage/testonly/admin_storage_tester.go (hunk 3 above)
Comment (codingllama) [reply storage/testonly/admin_storage_tester.go]: That's what happens on "TestHardDeleteTree" (apart from the freeze step). I'll add a frozen tree to the test.

File: storage/testonly/admin_storage_tester.go (hunk 2 above)
Comment (codingllama) [reply storage/testonly/admin_storage_tester.go]: There is a logic here: makeTree cares about softDelete, so I pushed it to the end of the file. A few other tests, like List and HardDelete care about soft-deleted trees too. OTOH, hardDelete and undelete are only called by their respective tests.

File: storage/mysql/admin_storage.go [hunk 4]
```diff
@@ -239,6 +245,14 @@ func readTree(row row) (*trillian.Tree, error) {
 	}
 	tree.PublicKey = &keyspb.PublicKey{Der: publicKey}
 
+	tree.Deleted = deleted.Valid && deleted.Bool
+	if deleteMillis.Valid {
```
[the remark this answers is not in the document]
Comment (codingllama) [reply storage/mysql/admin_storage.go]: Done.

**Comment by codingllama** [comment]
Comments addressed, let the Travis dance begin.

File: storage/testonly/admin_storage_tester.go (hunk 3 above)
Comment (daviddrysdale) [reply storage/testonly/admin_storage_tester.go]: Ah yes, I guess it does, hidden underneath the `makeTreeOrFail`  -- fine.

File: storage/testonly/admin_storage_tester.go (hunk 2 above)
Comment (daviddrysdale) [reply storage/testonly/admin_storage_tester.go]: Fair enough

**Pull request merged**